src/providers/aws.rs

```diff
@@ -0,0 +1,887 @@
+//! instances.
+//!
+//! # Example
+//! ```rust,no_run
+//! use tsunami::providers::{aws, Launcher};
+//! use tsunami::TsunamiBuilder;
+//!
+//! let mut b = TsunamiBuilder::default();
+//! b.add("my machine", aws::Setup::default()).unwrap();
+//! let mut l = aws::Launcher::default();
+//! b.spawn(&mut l).unwrap();
+//! let vms = l.connect_all().unwrap();
+//! let my_machine = vms.get("my machine").unwrap();
+//! let (stdout, stderr) = my_machine.ssh.as_ref().unwrap().cmd("echo \"Hello, EC2\"").unwrap();
+//! println!("{}", stdout);
+//! ```
+
+use crate::ssh;
+use crate::Machine;
+use educe::Educe;
+use failure::{Error, ResultExt};
+use itertools::Itertools;
+use rusoto_core::request::HttpClient;
+use rusoto_core::{DefaultCredentialsProvider, ProvideAwsCredentials, Region};
+use rusoto_ec2::Ec2;
+use std::collections::HashMap;
+use std::io::Write;
+use std::sync::Arc;
+use std::{thread, time};
+
+struct UbuntuAmi(String);
+
+impl From<Region> for UbuntuAmi {
+    fn from(r: Region) -> Self {
+        // https://cloud-images.ubuntu.com/locator/
+        // ec2 20190814 releases
+        UbuntuAmi(
+            match r {
+                Region::ApEast1 => "ami-e0ff8491",               // Hong Kong
+                Region::ApNortheast1 => "ami-0cb1c8cab7f5249b6", // Tokyo
+                Region::ApNortheast2 => "ami-081626bfb3fbc9f49", // Seoul
+                Region::ApSouth1 => "ami-0cf8402efdb171312",     // Mumbai
+                Region::ApSoutheast1 => "ami-099d318f80eab7e94", // Singapore
+                Region::ApSoutheast2 => "ami-08a648fb5cc86fb74", // Sydney
+                Region::CaCentral1 => "ami-0bc1dd4eb012a451e",   // Canada
+                Region::EuCentral1 => "ami-0cdab515472ca0bac",   // Frankfurt
+                Region::EuNorth1 => "ami-c37bf0bd",              // Stockholm
+                Region::EuWest1 => "ami-01cca82393e531118",      // Ireland
+                Region::EuWest2 => "ami-0a7c91b6616d113b1",      // London
+                Region::EuWest3 => "ami-033e0056c336ecff0",      // Paris
+                Region::SaEast1 => "ami-094c359b4d8c6a8ca",      // Sao Paulo
+                Region::UsEast1 => "ami-064a0193585662d74",      // N Virginia
+                Region::UsEast2 => "ami-021b7b04f1ac696c2",      // Ohio
+                Region::UsWest1 => "ami-056d04da775d124d7",      // N California
+                Region::UsWest2 => "ami-09a3d8a7177216dcf",      // Oregon
+                x => panic!("Unsupported Region {:?}", x),
+            }
+            .into(),
+        )
+    }
+}
+
+impl Into<String> for UbuntuAmi {
+    fn into(self) -> String {
+        self.0
+    }
+}
+
+/// Marker type for [`Setup`] indicating that it does not have an AMI.
+#[derive(Clone, Copy, Debug)]
+pub struct NoAmi;
+/// Marker type for [`Setup`] indicating that it has been initialized with an AMI.
+#[derive(Clone, Copy, Debug)]
+pub struct YesAmi;
+
+/// A descriptor for a particular machine setup in a tsunami.
```

Comment:
Maybe mention the fact that an AMI has to be set for this type to be used in association with `Launcher`?